components/raftstore/src/store/peer.rs

```diff
@@ -1764,57 +2019,156 @@ where
 
-            self.activate(ctx);
-            let mut meta = ctx.store_meta.lock().unwrap();
-            meta.readers
-                .insert(self.region_id, ReadDelegate::from_peer(self));
-        } else if has_msg {
-            let msgs = ready.take_persisted_messages();
-            self.send(ctx, msgs);
+                // Pause `read_progress` to prevent serving stale read while applying snapshot
+                self.read_progress.pause();
+            }
+            HandleReadyResult::NoIOTask { msgs } => {
+                if let Some(last) = self.unpersisted_readies.back_mut() {
+                    // Attach to the last unpersisted ready so that it can be considered to be
+                    // persisted with the last ready at the same time.
+                    if ready.number() <= last.max_empty_number {
+                        panic!(
+                            "{} ready number is not monotonically increaing, {} <= {}",
+                            self.tag,
+                            ready.number(),
+                            last.max_empty_number
+                        );
+                    }
+                    last.max_empty_number = ready.number();
+                    if !msgs.is_empty() {
+                        self.unpersisted_message_count += msgs.capacity();
+                        last.raft_msgs.push(msgs);
+                    }
+                } else {
+                    // If this ready don't need to be persisted and there is no previous unpersisted ready,
+                    // we can safely consider it is persisted so the persisted msgs can be sent immediately.
+                    self.persisted_number = ready.number();
+                    fail_point!("raft_before_follower_send");
+                    self.send_raft_messages(ctx, msgs);
+                    // The commit index and messages of light ready should be empty because no data needs
+                    // to be persisted. Only the committed entries may not be empty when the size is too
+                    // large to be fetched in the previous ready.
+                    let mut light_rd = self.raft_group.advance_append(ready);
+                    if let Some(idx) = light_rd.commit_index() {
+                        panic!(
+                            "{} advance ready that has no io task but commit index is changed to {}",
+                            self.tag, idx
+                        );
+                    }
+                    if !light_rd.messages().is_empty() {
+                        panic!(
+                            "{} advance ready that has no io task but message is not empty {:?}",
+                            self.tag,
+                            light_rd.messages()
+                        );
+                    }
+                    if !light_rd.committed_entries().is_empty() {
+                        self.handle_raft_committed_entries(ctx, light_rd.take_committed_entries());
```

Comment:
How can it happen?

Reply:
You can see the comments a few lines above. (now I move this comment above this line for clarity)